Add support for middlewares

jwfriese opened the pull request.

jwfriese committed:
Add support for middlewares

fortytw2 commented on handler.go:
```diff
@@ -111,6 +114,21 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+			}
+
+			h.errorHandler(w, err)
```
I believe ErrorHandler is supposed to write the actual error code.

What we should do here is refactor this to to use the internal exit code handler and not write headers twice (like this does)

jwfriese committed:
Allow error handler to write response status codes when middleware fails

jwfriese commented:
@fortytw2 - Addressed your comment, at least I think so. It leaves writing status codes to the error handler, which in the case of the DefaultErrorHandler, it already does this.
